### PR TITLE
SeqBatchScheduler with rolling batch

### PR DESCRIPTION
**Pull request by KexinFeng**
This PR succeeds #743 and implements contrastive search. Run `test_contrastive_scheduler` in `test_scheduler` to see the demo.

Here are the main features:

Iteration-level adding request and request existing, with padding and pad trimming
Enable resusable prefix kv_cache input for init_forward. The kv_cache is pre-calculated.
Greedy search, contrastive search

---
This PR is an iteration-level scheduler system, which is mainly implemented as a `class GreedySeqBatchScheduler()` and its parent class `SeqBatchScheduler`.  Another two important classes are: `SeqBatcher` and `Batch` (i.e. the `BatchList` in the java implementation)

See https://github.com/deepjavalibrary/djl/pull/2572 for detailed description of them.

To run this feature, see `test_scheduler.py`, which covers basically all the code. (Ignore the beam search and constrastive search part for now. They will be moved to a different PR)

Here are the main features:
1. Iteration-level adding request and request existing, with padding and pad trimming
2. Enable resusable prefix kv_cache input for init_forward. The kv_cache is pre-calculated.
3. Greedy search

Implementation directory
--- 
There are three major classes: `SeqBatchScheduler`, `SeqBatcher` and `Batch` and one major unit test: `test_greedy_scheduler` in this PR. See also https://github.com/deepjavalibrary/djl/pull/2572.

```python
"""
This is a scheduler, serving as an API to the consumer of the systme, allowing for three major actions: initForward, addBatch, fastForward, collectResults. 
"""
class SeqBatchScheduler:
    """
    This is a scheduler that manages the SeqBatcher, providing API which allows for actions like addBatch,
    collectResults.
    """

    def __init__(self, lm_block: LMBlock,
                 default_seq_batcher_cls: Type[SeqBatcher],
                 default_config: SearchConfig):
        self.default_search_configs = defaultdict(lambda: default_config)
        self.default_seq_batcher_cls = default_seq_batcher_cls
        self.lm_block = lm_block
        self.results: Dict[int, List[int]] = {}

        self.seq_batchers: Dict[
            Type[SeqBatcher]:List[SeqBatcher]] = defaultdict(
                list)  # {key: List[SeqBatcher]}

    def add_request(self,
                    input_ids: torch.Tensor,
                    request_uids: torch.Tensor,
                    seq_batcher_cls: Type[SeqBatcher] = None,
                    search_configs: List[SearchConfig] = None,
                    kv_cache: Union[Tuple, None] = None,
                    save_kv_cache_path: str = None):
    # Add requests tagged by request_uids


    def inference_call(self):
        """
        A sweep of inference calls on all seq_batchers in the scheduler
        Returns:
            output_ids (`Dict[Type[SeqBatcher]: List[List[int]]]`):
                About List[List[int]] structure, the outermost List[] corresponds to request_uid: List[int]. The
                inner List[int] is used to extend the output_ids sequences: past_output_ids: List[int].extend(List[
                int]). This is the same form as the output from @add_request.
            exist_request_uids (`Dict[Type[SeqBatcher]: List[int]]`):
                List[int] a list of request_uids that have finished.
        """
       pass
```

```python
class SeqBatcher(ABC):
    """
    This is a data class, which stores the search state (Batch), the control variables (eg seq_len, offsets, etc),
    and batch operations like merge_batch, trim, init_forward and inference_call. The latter two are search algorithm
    specific. Users may provide their own autoregressive searching algorithm by inheriting this class and overwriting
    the init_forward, inference_call along with the corresponding Batch.
    """

    def __init__(self, batch: Batch, request_uids: torch.Tensor,
                 offsets: torch.Tensor,
                 search_configs: defaultdict[Any,
                                             SearchConfig], lm_block: LMBlock):
        # Utility variables
        self.lm_block = lm_block
        self.exit_index = set()

        # Variables updated in a batch operation
        self.batch = batch
        self.request_uids = request_uids
        self.offsets = offsets
        self.search_configs = search_configs
        self.batch_size, _, self.seq_len, _ = batch.past_key_values[0][0].size(
        )

    @classmethod
    @abstractmethod
    def init_forward(
            cls,
            input_ids: torch.tensor,
            request_uids: torch.tensor,
            lm_block: LMBlock,
            search_configs: defaultdict[Any, SearchConfig],
            kv_cache: Union[Tuple, None] = None,
            kv_cache_input_ids: Union[torch.tensor, None] = None,
            save_kv_cache_path=None) -> Tuple[SeqBatcher, List[List[str]]]:
        pass

    @abstractmethod
    def forward(self) -> List[List[int]]:
        pass

    @torch.no_grad()
    def add_batch(self, seq_batcher: SeqBatcher):
        pass

    @torch.no_grad()
    def collect_and_trim(self) -> None:
        pass

    def exit_criteria(self, output_ids: torch.Tensor, search_configs):
        pass

```

```python
"""
Batch is a data class consisting of fields of tensors like past_output_ids, past_key_values,  past_attention_mask. It's a compact collection of variables that need to be updated in each incremental inference call.
"""
class Batch:

    def __init__(self,
                 past_output_ids: torch.Tensor = None,
                 past_attention_mask: torch.Tensor = None,
                 past_key_values=None,
                 logits: torch.Tensor = None):
        self.past_output_ids = past_output_ids
        self.past_key_values = past_key_values
        self.past_attention_mask = past_attention_mask
        self.logits = logits

    # merges another batch with itself.
    def merge(self, batch: Batch, seq_delta: int) -> Batch:

    def trim(self, keep_indices: torch.Tensor, trim_seq_len: int):

```

Miscellany
---
`assemble_prefix_kv_cache` and `nudge_to_squeeze_bubble_padding` are used only when a prefix kv_cache is provided. They are small augmented feature.

### CODE REVIEW

File: engines/python/setup/djl_python/scheduler/__init__.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS"
+# BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+
+from .lm_block import HuggingfaceBlock
+from .seq_batch_scheduler_impl import GreedySeqBatchScheduler
+from .search_config import SearchConfig
+#!/usr/bin/env python
```
Comment (lanking520) [inline engines/python/setup/djl_python/scheduler/__init__.py:17]: need clean up

**Review by lanking520 — APPROVED** [review]
